metric_providers/disk/io/cgroup/container/source.c

```diff
@@ -0,0 +1,189 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <errno.h>
+#include <unistd.h>
+#include <sys/time.h>
+#include <string.h> // for strtok
+#include <getopt.h>
+
+typedef struct container_t { // struct is a specification and this static makes no sense here
+    char path[BUFSIZ];
+    char *id;
+} container_t;
+
+typedef struct disk_io_t { // struct is a specification and this static makes no sense here
+    long int rbytes;
+    long int wbytes;
+} disk_io_t;
+
+
+// All variables are made static, because we believe that this will
+// keep them local in scope to the file and not make them persist in state
+// between Threads.
+// in any case, none of these variables should change between threads
+static int user_id = 0;
+static unsigned int msleep_time=1000;
+
+
+static disk_io_t get_disk_cgroup(char* filename) {
+    long int rbytes = -1;
+    long int rbytes_sum = 0;
+    long int wbytes = -1;
+    long int wbytes_sum = 0;
+
+    FILE * fd = fopen(filename, "r");
+    if ( fd == NULL) {
+        fprintf(stderr, "Error - Could not open path for reading: %s. Maybe the container is not running anymore? Are you using --rootless mode? Errno: %d\n", filename, errno);
+        exit(1);
+    }
+
+    while (fscanf(fd, "%*u:%*u rbytes=%ld wbytes=%ld rios=%*u wios=%*u dbytes=%*u dios=%*u", &rbytes, &wbytes) == 2) {
+        rbytes_sum += rbytes;
+        wbytes_sum += wbytes;
+    }
+    ;
+
+    fclose(fd);
+
+    if(rbytes < 0 || wbytes < 0) {
+        fprintf(stderr, "Error - io.stat could not be read or was < 0.");
+        exit(1);
+    }
+
+    disk_io_t disk_io;
+    disk_io.rbytes = rbytes_sum;
+    disk_io.wbytes = wbytes_sum;
+    return disk_io;
+}
+
+static int output_stats(container_t *containers, int length) {
+
+    struct timeval now;
+    int i;
+
+    gettimeofday(&now, NULL);
+    for(i=0; i<length; i++) {
+        disk_io_t disk_io = get_disk_cgroup(containers[i].path);
+        printf("%ld%06ld %ld %s\n", now.tv_sec, now.tv_usec, disk_io.rbytes+disk_io.wbytes, containers[i].id);
+    }
+    usleep(msleep_time*1000);
+
+    return 1;
+}
+
+static int parse_containers(container_t** containers, char* containers_string, int rootless_mode) {
+    if(containers_string == NULL) {
+        fprintf(stderr, "Please supply at least one container id with -s XXXX\n");
+        exit(1);
+    }
+
+    *containers = malloc(sizeof(container_t));
+    char *id = strtok(containers_string,",");
+    int length = 0;
+
+    for (; id != NULL; id = strtok(NULL, ",")) {
+        //printf("Token: %s\n", id);
+        length++;
+        *containers = realloc(*containers, length * sizeof(container_t));
+        (*containers)[length-1].id = id;
```

Comment:
I would copy the id here instead of referencing a pointer. You don't know what will happen with containers_string in the future

Also I would trim the string as someone could specify container_string:

```"a, b"```

Reply:
Yeah, I agree copying is here the better choice. Especially since strtok anyway has modified the string it should be considered no-reuse memory.

Also added some more NUL-Byte terminations which where missing.

2. I would not trim the string. This is a 5-10 line function in C and I would rather consider this a user-submitted-data error. It will fail cause the container cannot be found.